vsphere-syncer container panic: runtime error: invalid memory address or nil pointer dereference

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug


**What happened**:
Upgraded vsphere-csi-driver from 2.7.1 to 3.0.1 and the new version is unable to provision new block PVs.

**How to reproduce it (as minimally and precisely as possible)**:
Followed the steps to upgrade:
https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-3F277B52-68CC-4125-AD0F-E7293940B4B4.html

**Anything else we need to know?**:
`vsphere-csi-controller/vsphere-syncer` container crashes the the following error message when a new PVC is created in the cluster.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x1b3f1e8]

goroutine 191 [running]:
sigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.csiGetPVtoBackingDiskObjectIdMapping({0x26c5318, 0xc0002c71d0}, {0x26dd320, 0xc00080b980}, 0xc000176000)
	/build/pkg/syncer/pv_to_backingdiskobjectid_mapping.go:50 +0x168
sigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.InitMetadataSyncer.func12()
	/build/pkg/syncer/metadatasyncer.go:601 +0xd9
created by sigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.InitMetadataSyncer
	/build/pkg/syncer/metadatasyncer.go:597 +0x254c
```

Full logs: 
[vsphere-syncer.log.tar.gz](https://github.com/kubernetes-sigs/vsphere-csi-driver/files/11416074/vsphere-syncer.log.tar.gz)


**Environment**:
- container versions:
```
k8s.gcr.io/sig-storage/csi-attacher:v4.2.0
k8s.gcr.io/sig-storage/csi-provisioner:v3.4.0
k8s.gcr.io/sig-storage/csi-resizer:v1.7.0
k8s.gcr.io/sig-storage/csi-snapshotter:v6.2.1
k8s.gcr.io/sig-storage/livenessprobe:v2.9.0
gcr.io/cloud-provider-vsphere/csi/release/driver:v3.0.1
gcr.io/cloud-provider-vsphere/csi/release/syncer:v3.0.1
```
- Kubernetes version: 1.25.6
- vSphere version: 7.0.3.01200
- OS (e.g. from /etc/os-release): Ubuntu 22.04 LTS
- Kernel (e.g. `uname -a`): kernel 5.15.0-71-generic

--- a/pkg/syncer/metadatasyncer.go
+++ b/pkg/syncer/metadatasyncer.go
@@ -587,18 +587,55 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
 		defer pvToBackingDiskObjectIdMappingTicker.Stop()
 
 		var pvToBackingDiskObjectIdSupportCheck bool
-		vCenter, err := cnsvsphere.GetVirtualCenterInstance(ctx, configInfo, false)
-		if err != nil {
-			return err
+		if !isMultiVCenterFssEnabled {
+			vCenter, err := cnsvsphere.GetVirtualCenterInstance(ctx, configInfo, false)
+			if err != nil {
+				return err
+			}
+			pvToBackingDiskObjectIdSupportCheck = common.CheckPVtoBackingDiskObjectIdSupport(ctx, vCenter)
+
+		} else {
+			vcconfigs, err := cnsvsphere.GetVirtualCenterConfigs(ctx, configInfo.Cfg)
+			if err != nil {
+				return logger.LogNewErrorf(log, "failed to get VirtualCenterConfigs. err: %v", err)
+			}
+			var pvToBackingDiskObjectIdSupportedByVc bool
+			for _, vcconfig := range vcconfigs {
+				vCenter, err := cnsvsphere.GetVirtualCenterInstanceForVCenterConfig(ctx, vcconfig, false)
+				if err != nil {
+					return logger.LogNewErrorf(log, "failed to get vCenterInstance for vCenter Host: %q, err: %v", vcconfig.Host, err)
+				}
+				// All VCs in the deployment must support PV to Backing Disk Object ID feature.
+				pvToBackingDiskObjectIdSupportedByVc = common.CheckPVtoBackingDiskObjectIdSupport(ctx, vCenter)
+				if !pvToBackingDiskObjectIdSupportedByVc {
+					pvToBackingDiskObjectIdSupportCheck = false
+					break
+				} else {
+					pvToBackingDiskObjectIdSupportCheck = true
+				}
+			}
 		}
-		pvToBackingDiskObjectIdSupportCheck = common.CheckPVtoBackingDiskObjectIdSupport(ctx, vCenter)
 
 		if pvToBackingDiskObjectIdSupportCheck {
 			go func() {
 				for ; true; <-pvToBackingDiskObjectIdMappingTicker.C {
 					ctx, log = logger.GetNewContextWithLogger()
 					log.Info("get pv to backingDiskObjectId mapping is triggered")
-					csiGetPVtoBackingDiskObjectIdMapping(ctx, k8sClient, metadataSyncer)
+
+					if !isMultiVCenterFssEnabled {
+						csiGetPVtoBackingDiskObjectIdMapping(ctx, k8sClient, metadataSyncer,
+							metadataSyncer.configInfo.Cfg.Global.VCenterIP)
+					} else {
+						vcconfigs, err := cnsvsphere.GetVirtualCenterConfigs(ctx, configInfo.Cfg)
+						if err != nil {
+							log.Error(log, "failed to get VirtualCenterConfigs. err: %v", err)
+							return
+						}
+						// Update mapping for all VCs.
+						for _, vcconfig := range vcconfigs {
+							csiGetPVtoBackingDiskObjectIdMapping(ctx, k8sClient, metadataSyncer, vcconfig.Host)
+						}
+					}
 				}
 			}()
 		}

--- a/pkg/syncer/pv_to_backingdiskobjectid_mapping.go
+++ b/pkg/syncer/pv_to_backingdiskobjectid_mapping.go
@@ -31,9 +31,9 @@ import (
 
 // TODO: refactor pv to backingdiskobjectid and volume health code to reduce duplicated code
 func csiGetPVtoBackingDiskObjectIdMapping(ctx context.Context, k8sclient clientset.Interface,
-	metadataSyncer *metadataSyncInformer) {
+	metadataSyncer *metadataSyncInformer, vc string) {
 	log := logger.GetLogger(ctx)
-	log.Debug("csiGetPVtoBackingDiskObjectIdMapping: start")
+	log.Debugf("csiGetPVtoBackingDiskObjectIdMapping for %s: start", vc)
 	// Call CNS QueryAll to get container volumes by cluster ID.
 	queryFilter := cnstypes.CnsQueryFilter{
 		ContainerClusterIds: []string{
@@ -47,16 +47,23 @@ func csiGetPVtoBackingDiskObjectIdMapping(ctx context.Context, k8sclient clients
 			string(cnstypes.QuerySelectionNameTypeVolumeType),
 		},
 	}
-	queryAllResult, err := metadataSyncer.volumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+
+	cnsVolumeMgr, err := getVolManagerForVcHost(ctx, vc, metadataSyncer)
 	if err != nil {
-		log.Errorf("csiGetPVtoBackingDiskObjectIdMapping: failed to QueryAllVolume with err=%+v", err.Error())
+		log.Errorf("csiGetPVtoBackingDiskObjectIdMapping for %s: Failed to get volume manager. Err: %v", vc, err)
+		return
+	}
+
+	queryAllResult, err := cnsVolumeMgr.QueryAllVolume(ctx, queryFilter, querySelection)
+	if err != nil {
+		log.Errorf("csiGetPVtoBackingDiskObjectIdMapping for %s: failed to QueryAllVolume with err=%+v", vc, err)
 		return
 	}
 
 	// Get K8s PVs in State "Bound".
 	k8sPVs, err := getBoundPVs(ctx, metadataSyncer)
 	if err != nil {
-		log.Errorf("csiGetPVtoBackingDiskObjectIdMapping: Failed to get PVs from kubernetes. Err: %+v", err)
+		log.Errorf("csiGetPVtoBackingDiskObjectIdMapping for %s: Failed to get PVs from kubernetes. Err: %+v", vc, err)
 		return
 	}
 
@@ -71,13 +78,13 @@ func csiGetPVtoBackingDiskObjectIdMapping(ctx context.Context, k8sclient clients
 			pvc, err := metadataSyncer.pvcLister.PersistentVolumeClaims(
 				pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
 			if err != nil {
-				log.Warnf("csiGetPVtoBackingDiskObjectIdMapping: Failed to get pvc for namespace %s and name %s. err=%+v",
-					pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name, err)
+				log.Warnf("csiGetPVtoBackingDiskObjectIdMapping for %s: Failed to get pvc for namespace %s and name %s. err=%+v",
+					vc, pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name, err)
 				continue
 			}
 			volumeHandleToPvcMap[pv.Spec.CSI.VolumeHandle] = pvc
-			log.Debugf("csiGetPVtoBackingDiskObjectIdMapping: pvc %s/%s is backed by pv %s volumeHandle %s, pv uid is %s",
-				pvc.Namespace, pvc.Name, pv.Name, pv.Spec.CSI.VolumeHandle, string(pv.UID))
+			log.Debugf("csiGetPVtoBackingDiskObjectIdMapping for %s: pvc %s/%s is backed by pv %s volumeHandle %s, pv uid is %s",
+				vc, pvc.Namespace, pvc.Name, pv.Name, pv.Spec.CSI.VolumeHandle, string(pv.UID))
 			volumeHandleToPvUidMap[pv.Spec.CSI.VolumeHandle] = string(pv.UID)
 		}
 	}
@@ -98,19 +105,27 @@ func csiGetPVtoBackingDiskObjectIdMapping(ctx context.Context, k8sclient clients
 	}
 
 	for volID, pvc := range volumeHandleToPvcMap {
+
+		if _, ok := volumeIdToBackingObjectIdMap[volID]; !ok {
+			log.Debugf("csiGetPVtoBackingDiskObjectIdMapping for %s: Skipping vol %s as it is not found on this VC.",
+				vc, volID)
+			continue
+		}
+
 		pvToBackingDiskObjectIdPair := volumeHandleToPvUidMap[volID] + ":" + volumeIdToBackingObjectIdMap[volID]
 		updated, err := updatePVtoBackingDiskObjectIdMappingStatus(ctx, k8sclient, pvc, pvToBackingDiskObjectIdPair)
 		if err != nil {
-			log.Error("csiGetPVtoBackingDiskObjectIdMapping: Failed to update pv to backingDiskObjectId mapping")
+			log.Error("csiGetPVtoBackingDiskObjectIdMapping for %s: Failed to update pv to backingDiskObjectId mapping",
+				vc)
 			return
 		}
 		if updated {
-			log.Infof("csiGetPVtoBackingDiskObjectIdMapping: pvc %s is updated with pv to backingDiskObjectId mapping %s",
-				pvc.Name, pvToBackingDiskObjectIdPair)
+			log.Infof("csiGetPVtoBackingDiskObjectIdMapping for %s: pvc %s is updated with pv to backingDiskObjectId mapping %s",
+				vc, pvc.Name, pvToBackingDiskObjectIdPair)
 		}
 	}
 
-	log.Debug("csiGetPVtoBackingDiskObjectIdMapping: end")
+	log.Debugf("csiGetPVtoBackingDiskObjectIdMapping for %s: end", vc)
 }
 
 func validatePvToBackingDiskObjectIdPair(pvToBackingDiskObjectIdPair string) bool {

